cpovc_ovc/functions.py

```diff
@@ -631,3 +637,121 @@ def save_viral_load(request):
+
+    @method_once
+    def generate_token(self):
+        # generate token.
+        login_url = self.login_url
+        headers = {'Content-Type': 'application/x-www-form-urlencoded'}
+        auth = (self.client_id, self.client_secret)
+        credentials = { "grant_type": self.grant_type, 
+                        "username": self.username, 
+                        "password": self.password, 
+                        "scope": self.scope }
+        response = requests.post(login_url, headers=headers, data=credentials, auth=auth)
+        if response.status_code == 200:
+            json_token = json.loads(response.content)
+            api_token = json_token.get('access_token')
+            return api_token
+        else:
+            print(response.content)
+    
+
+    def get_facilities(self):
+        # request for facilities
+        headers = {'Authorization': 'Bearer {0}'.format(self.api_token)}
+        api_url = '{0}facilities/facilities/?format=json'.format(self.api_base_url)
+        response = requests.get(api_url, headers=headers)
+        
+        if response.status_code == 200:
+            json_object = json.loads(response.content)
+            return json_object
+        else:
+            print(response.content, api_url)
+
+
+    def get_subcounty_id(self, facility_id):
+        # request facility and get sub-county id.
+        headers = {'Authorization': 'Bearer {0}'.format(self.api_token)}
+        payload = {'format': 'json'}
+        api_url = '{0}{1}'.format(self.facility_base_url, facility_id)
+        response = requests.get(api_url, headers=headers, params=payload)
+        if response.status_code == 200:
+            json_object = json.loads(response.content)
+            cpims_subcounty_id = json_object["constituency_code"]
+            return cpims_subcounty_id
+        else:
+            print(response, api_url)
+
+    @transaction.atomic
+    def get_newest_facilities(self):
+        # loop for new facilities.
+        try:
+            data = self.get_facilities()
+            results = data["results"]
+            for facility in results:
+                facility_code = facility["code"]
+                if facility_code > self.latest_facility:
+                    facility_id = facility["id"]
+                    cpims_subcounty_id = self.get_subcounty_id(facility_id)
+                    facility_name = facility["official_name"]
+                    sub_county_id = facility["sub_county_id"]
+                    new_facility = OVCFacility(facility_code=facility_code, 
+                        facility_name=facility_name, 
+                        sub_county_id=cpims_subcounty_id)
+                    
+                    new_facility.save()
+        
+        except Exception as e:
+            raise e
+        else:
+            pass
+
+    def schedule_update(self):
+        # Update facilities every tuesday at 10:15 am
+        schedule.every().tuesday.at("10:15").do(self.get_newest_facilities)
```

Comment:
Facility updates via api are preferred done between midnight nd and 6am

Reply:
Fixed bd80b0f